File: go/src/BronKerbosch/bron_kerbosch3_gpx.go
package BronKerbosch

func bronKerbosch3gpx(graph *UndirectedGraph, reporter Reporter) {
	// Bron-Kerbosch algorithm with degeneracy ordering, with nested searches
	// choosing a pivot from both candidates and excluded vertices (IK_GPX).
	var ordering SimpleVertexVisitor
	degeneracyOrdering(graph, &ordering, -1)
	excluded := make(VertexSet, len(ordering.vertices))
	for _, v := range ordering.vertices {
		neighbours := graph.neighbours(v)
		neighbouringCandidates := neighbours.Difference(excluded)
		if !neighbouringCandidates.IsEmpty() {
			neighbouringExcluded := neighbours.Intersection(excluded)
			visit(
				graph, reporter,
				MaxDegree, MaxDegreeLocalX,
				neighbouringCandidates,
				neighbouringExcluded,
				[]Vertex{v})
		}
		excluded.Add(v)
	}
}

File: rust/bron_kerbosch/src/core/bron_kerbosch3_gpx.rs
//! Bron-Kerbosch algorithm with degeneracy ordering, with nested searches
//! choosing a pivot from both candidates and excluded vertices (IK_GPX)

use super::bron_kerbosch_degen::explore_with_degeneracy;
use crate::core::bron_kerbosch_pivot::PivotChoice;
use crate::{BronKerboschAlgorithm, CliqueAccumulator, Graph, VertexSetLike};

pub struct Algo();
impl BronKerboschAlgorithm for Algo {
    fn name() -> String {
        String::from("Ver3½-GPX")
    }

    fn explore<VertexSet, Accumulator>(
        graph: &Graph<VertexSet>,
        min_clique_size: usize,
        accumulator: Accumulator,
    ) -> Accumulator::Harvest
    where
        VertexSet: VertexSetLike,
        Accumulator: CliqueAccumulator,
    {
        explore_with_degeneracy(
            graph,
            min_clique_size,
            accumulator,
            PivotChoice::MaxDegreeLocalX,
        )
    }
}
